FWCore/Framework/interface/EventSetupRecord.h

```diff
@@ -207,26 +205,32 @@ namespace edm {
     protected:
       template <template <typename> typename H, typename T, typename R>
       H<T> getHandleImpl(ESGetToken<T, R> const& iToken) const {
-        if
-          UNLIKELY(iToken.transitionID() != transitionID()) { throwWrongTransitionID(); }
-        assert(iToken.isInitialized());
+        if UNLIKELY (not iToken.isInitialized()) {
+          std::rethrow_exception(makeInvalidTokenException(this->key(), DataKey::makeTypeTag<T>()));
+        }
+        if UNLIKELY (iToken.transitionID() != transitionID()) {
+          throwWrongTransitionID();
+        }
         assert(getTokenIndices_);
         //need to check token has valid index
-        if
-          UNLIKELY(not iToken.hasValidIndex()) { return invalidTokenHandle<H>(iToken); }
+        if UNLIKELY (not iToken.hasValidIndex()) {
+          return invalidTokenHandle<H>(iToken);
+        }
```

Comment:
... if this check (which is the only caller of `makeInvalidTokenException()`) would still be useful? Can `ESGetToken` be constructed with a valid transition ID but invalid index?

Reply:
If the data product is not available that would happen

Reply:
Then the exception message of that appears to be confusing. `invalidTokenHandle()`
https://github.com/cms-sw/cmssw/blob/cea1d1bc4aa7b376b68eb70a34cfb4c7184f438b/FWCore/Framework/interface/EventSetupRecord.h#L252-L256
just leads to `makeInvalidTokenException()` that creates
https://github.com/cms-sw/cmssw/blob/cea1d1bc4aa7b376b68eb70a34cfb4c7184f438b/FWCore/Framework/src/EventSetupRecord.cc#L73-L81
that suggests to call consumes, which should be unrelated to (in)availability of the data product.

Reply:
We concluded that a valid transition ID and invalid index should not happen, so I'll replace this exception message.